uiowa/uiowa
# Now News Migrations

Relates to #5887 




# How to test


Start up Dev Desktop and sync down a fresh copy of now.uiowa.edu (on uiowa 703).
`git checkout now-news-migrations && git pull`
`ddev composer install` if you haven't recently
`ddev blt ds --site=now.uiowa.edu`
(Syncing will possibly fail on mismatched config. the `pmu && en` below will import the new config and fix the issue, and testing can continue as normal)
```
ddev drush @now.local pmu now_migrate && ddev drush @now.local en now_migrate && ddev drush @now.local uli /admin/config/sitenow/sitenow-migrate
```
Complete the form following the fields' help texts
`drush @now.local ms` <- check that the migrations are registered and listed
You can run each migration individually using their migration ids (eg `drush @now.local mim now_achievement`), or run them all using `drush @now.local mim --all`
You can roll back an individual migration using `drush @now.local mr migration_id` (this will not remove created media entities from that migration)
You can limit a migration to X number of entities using the `--limit=X` option. (This can be particularly useful for the news feature migration. I was able to run the full thing without a memory issue, but it does slow down over time. Batches of `drush @now.local mim now_news_feature --limit=1000` may help)

# Things to check for all news migrations
Dates and Summaries should look alright (https://now.uiowa.ddev.site/news). 
General spotchecking
Images/files within body text should have been brought over and inserted, with roughly matching styling (it's not always one-to-one, but things shouldn't be too out of place)
Tags taxonomy should have many new tags created as part of the migration, but they should not be duplicative
Some errors while running the migration are expected to be seen in the terminal output (eg ` [error]  Invalid marker found at offset 0: 0x47 `), but these should not stop the migration from proceeding. Feel free to document any errors encountered to double check, though.

## Achievements migration

Article type should be set to Achievement
The original byline should be set as the Article Source
Primary Media should be pulled into the featured image, but default them to "do not display" (should only be used for meta/teasers)
(Field Caption??)
Preserve Links message should match existing site (`select * from node__field_article_preserve_links where field_article_preserve_links_value = 1;` is an example SQL query to find node ids for which it is turned on post-migration)
Achievement Category, News From, News For, News About, and Keywords should all be brought over as Tags (along with an additional 'achievement' tag to match Article Type)

## News Feature migration

Article type should be set to News Feature
Subhead should be brought over and prepended to the body with the Intro Light style
The original byline should be set as the Article Source
Preserve Links message should match existing site (`select * from node__field_article_preserve_links where field_article_preserve_links_value = 1;` is an example SQL query to find node ids for which it is turned on post-migration)
Primary Media should be pulled into the featured image, but default them to "do not display" (should only be used for meta/teasers)
News From, News About, News For, and Keywords should be pulled into Tags
The photo gallery should be present (`select distinct entity_id from node__field_gallery_images;` sql query can get you the node ids of any articles that have a photo gallery post-migration)
Related Items -> Related Content

### Embargo Information
At the time of this writing, there is one article that has embargo information included in its field, and should be checked that 1) the embargo information has pulled over correctly and 2) that the node is not published
`select * from node__field_embargo_information;` is a simple SQL query to check for any nodes with embargo info on the new site post-migration, and the resulting entity_id can be used to check the node itself

### Videos

Old articles had the option of video content as their Primary Media. Articles with that originally should now have that video turned into an oembed (remote video) media entity and embedded at the top of the body.
Quick way to check these is to go into https://now.uiowa.ddev.site/admin/content/media, filter by Remote Video, and check the usage for any of them--that should include the article to which they've been attached

## In the News migration
Article type should be In the News unless the article source is a uiowa.edu domain--then it should be UI Spotlight
Source name/link should map to Org/Url
Article should link directly to source
Original publication date should be updated correctly
News From, News About, News For, and Keywords should be pulled into Tags

# Updates from dev run
Inline videos
Originally, inline embedded videos were not brought over (ex: https://now.uiowa.edu/2016/08/holden-comprehensive-cancer-center-ui-receives-highest-designation-nci). This has since been updated, and both News Feature and Achievement migrations should properly create inline embedded video media

Featured images
On News Feature migration, featured images are brought over and, if they are a 3:2 or wider aspect ratio, set to display else set not to display.

Image (and video) captions
Captions were not originally styled/connected to their video and instead displayed as plain `<p>` tags in the content. They should now properly be attached to their respective image/video

Extra spacing in captions
Similar to captions not being attached to their respective image/video, originally there were large amounts of additional space added using single or multiple `<br>` tags

Callouts
Originally callouts did not have a direct mapping. They have been updated (halfway there--awaiting full component availability)

Gallery Images missing captions
Gallery images should now have a global caption on their respective images which may or may not match the image alt or title tags. These should match the source gallery image captions

News Feature byline update
Bylines from source were added to the Article Source Organization field in the destination. Since the initial dev run, this has been updated to prepend it with "Written by: " to more closely match styling of referenced authors
